packages/react-form-hooks/src/validation.ts

```diff
@@ -0,0 +1,47 @@
+interface Matcher<Input, ValidationContext extends object = {}> {
+  (input: Input, context: ValidationContext): boolean;
+}
+
+interface ErrorContentFunction<Input> {
+  (input: Input): string;
+}
+
+type ErrorContent<Input> = string | ErrorContentFunction<Input>;
+
+export function validate<Input = string, ValidationContext extends object = {}>(
+  matcher: Matcher<Input, ValidationContext>,
+) {
+  return (errorContent: ErrorContent<Input>) => {
+    return (input: Input, context: ValidationContext) => {
+      const matches = matcher(input, context);
+
+      if (matches) {
+        return;
+      }
+
+      if (typeof errorContent === 'function') {
+        return errorContent(input);
+      }
+
+      return errorContent;
+    };
+  };
+}
+
+export function lengthMoreThan(length: number, error: ErrorContent<string>) {
+  return validate(input => {
+    return input.length > length;
+  })(error);
+}
+
+export function notEmpty(error: ErrorContent<string>) {
```

Comment:
The API for `validate` is slightly different that form-state's. The new API let's you more easily create reusable validators, since there's no real purpose to using `validate` for one offs (it's almost always easier to just pass a function directly to the field).

```
// old way
validate(matcher, errorCreator);

// eg
const title = useField(() => {
  value: '',
  validates: validate((value) => value.includes('✨'), 'Title must include sparkles!')
});

// new way
const validator = validate(matcher)
const error = validator(errorCreator)

// eg
const hasSpakles = validate((value) => value.includes('✨'))

const title = useField({
  value: '',
  validates: hasSparkles("Titles must have sparkles"),
});
```